code/modules/power/smes.dm

```diff
@@ -153,12 +157,38 @@
 	if(terminal && terminal.powernet)
 		inputted_power = terminal.powernet.draw_power(inputted_power)
 		charge += inputted_power * SMESRATE
+		input_taken = inputted_power
 		if(percentage == 100)
 			inputting = 2
 		else if(percentage)
 			inputting = 1
 		// else inputting = 0, as set in process()
 
+/obj/machinery/power/smes/proc/update_time()

```

Comment:
As discussed in Discord last night, this method of calculation:

1. Seems really complex.
2. Doesn't reflect the actual nature of time. (Though it serves as a good enough approximation I _guess_ until someone changes how often machinery ticks.)

A more simple formula would be:
```dm
	var/delta_power = input_taken - output_used
	delta_power *= SMESRATE

	var/goal = (delta_power < 0 ? charge : (capacity - charge))
	// you can use the sign of delta_power here to set charge_mode

	var/time_secs = (goal / abs(delta_power)) * 2
```

Reply:
Okay, I have reworked it. Also changed 2 to specific time passed between proc was called. Also added sanity check if delta_power is 0, so that no division by 0